database/scrapers/event_calendar.ts

```diff
@@ -0,0 +1,227 @@
+import { DateTime } from "luxon";
+
+import { BaseScraperModule, TaskHandle } from "#commands/db_scrape";
+import CalendarEvent from "#models/calendar_event";
+import { ICSObject, ICSRoot, ICSValue, parseICS } from "#utils/ics";
+
+const FETCH_LIMIT_DAYS = 30;
+const CALENDAR_ID = "9ke30hbjjke60u5jbii42g2rpo@group.calendar.google.com";
+const FIELD_LABELS = {
+  startDate: "DTSTART",
+  endDate: "DTEND",
+  googleCalId: "UID",
+  desc: "DESCRIPTION",
+  location: "LOCATION",
+  name: "SUMMARY",
+};
+
+interface GoogleCalendarEventDto {
+  googleCalId: string;
+  startTime: DateTime<true>;
+  endTime: DateTime<true>;
+  name: string;
+  location: string | null;
+  description: string | null;
+}
+
+class CalendarParser {
+  private events: ICSObject[];
+
+  constructor(
+    private calendarString: string,
+    private task: TaskHandle,
+    logger: (message: string) => void = (message) => this.task.update(message),
+  ) {
+    if (
+      !this.calendarString.startsWith("BEGIN:VCALENDAR") ||
+      !this.calendarString.trimEnd().endsWith("END:VCALENDAR")
+    ) {
+      throw new Error("Calendar invalid");
+    }
+    const calendarRoot = parseICS(calendarString, logger).VCALENDAR;
+    if (calendarRoot === undefined) {
+      throw new Error("Calendar invalid - missing calendar root");
+    }
+    const eventArray = (calendarRoot as ICSRoot).VEVENT;
+    if (eventArray === undefined || !Array.isArray(eventArray)) {
+      throw new Error("Calendar invalid - missing event array");
+    }
+    this.events = eventArray;
+  }
+
+  public *getLastEvent(): Generator<GoogleCalendarEventDto, void, undefined> {
+    const validEvents = this.events
+      .map((e) => this.extractEvent(e))
+      .filter((e) => e !== undefined)
+      .sort((a, b) => {
+        const diff = a.endTime.diff(b.endTime).milliseconds;
+        if (diff !== 0) {
+          return diff;
+        }
+        return a.startTime.diff(b.startTime).milliseconds;
+      }); //sorted from earliest to last, with one most recent/in the future being last
+    for (let i = validEvents.length - 1; i >= 0; i--) {
+      yield validEvents[i];
+    }
+  }
+
+  //null - none left, undefined - malformed event
+  private extractEvent(
+    eventBlock: ICSObject,
+  ): GoogleCalendarEventDto | undefined {
+    if (Array.isArray(eventBlock)) {
+      this.task.update(
+        "Invalid calendar event - block is not an ICSObject. Skipping...",
+      );
+      return undefined;
+    }
```

Comment:
surprising that the linter doesn't pick this up as a condition that can never be met, since it's known to be an `ICSObject`, not an array.